src/FsAutoComplete.Core/TestServer.fs

```diff
@@ -0,0 +1,223 @@
+
+            if m.Success then
+                let processId = m.Groups.[1].Value
+                Some processId
+            else
+                None
+
+        member val TestResults : TestResult ResizeArray = ResizeArray() with get,set
+
+        interface ITestRunEventsHandler with
+            member _.HandleLogMessage (_level: TestMessageLevel, message: string): unit = 
+                match tryGetDebugProcessId message with
+                | Some processId -> notifyIncrementalUpdate (ProcessWaitingForDebugger processId)
+                | None -> () 
+
+            member _.HandleRawMessage (_rawMessage: string): unit = 
+                ()
+
+            member this.HandleTestRunComplete (_testRunCompleteArgs: TestRunCompleteEventArgs, lastChunkArgs: TestRunChangedEventArgs, _runContextAttachments: System.Collections.Generic.ICollection<AttachmentSet>, _executorUris: System.Collections.Generic.ICollection<string>): unit = 
+                if((not << isNull) lastChunkArgs && (not << isNull) lastChunkArgs.NewTestResults) then
+                    this.TestResults.AddRange(lastChunkArgs.NewTestResults)
+                    notifyIncrementalUpdate (Progress lastChunkArgs)
+
+            member this.HandleTestRunStatsChange (testRunChangedArgs: TestRunChangedEventArgs): unit = 
+                if((not << isNull) testRunChangedArgs && (not << isNull) testRunChangedArgs.NewTestResults) then
+                    this.TestResults.AddRange(testRunChangedArgs.NewTestResults)
+                    notifyIncrementalUpdate (Progress testRunChangedArgs)
+
+            member _.LaunchProcessWithDebuggerAttached (_testProcessStartInfo: TestProcessStartInfo): int = 
+                raise (System.NotImplementedException())
+
+    module TestPlatformOptions = 
+        let withTestCaseFilter (options: TestPlatformOptions) filterExpression =
+            options.TestCaseFilter <- filterExpression 
+    
+
+    let runTestsAsync (vstestPath: string) (incrementalUpdateHandler: TestRunUpdate -> unit) (sources: TestProjectDll list) (testCaseFilter: string option) (shouldDebug: bool) : Async<TestResult list> = 
+        async {
+            let consoleParams = ConsoleParameters()
+            if shouldDebug then
+                consoleParams.EnvironmentVariables <- [
+                    "VSTEST_HOST_DEBUG", "1"
+                ] |> dict |> System.Collections.Generic.Dictionary
+            let vstest = new VsTestConsoleWrapper(vstestPath, consoleParams)
+            let runHandler = TestRunHandler(incrementalUpdateHandler)
+            
+            let options = new TestPlatformOptions()
+            testCaseFilter |> Option.iter (TestPlatformOptions.withTestCaseFilter options)
+            
+            use! _cancel = Async.OnCancel(fun () -> 
+                printfn "Cancelling test run"
+                vstest.CancelTestRun()
+                printfn "Test Run Cancelled")
+
+            vstest.RunTests(sources, null, options, runHandler) 
```

Comment:
@nohwnd is the best to help with VSTest stuff. But, I see there is a `RunTestsWithCustomTestHost` API on `VsTestConsoleWrapper`. From there, you have `LaunchTestHost` method with `TestProcessStartInfo` parameter, so that you can start the process yourself and then you know its id. That should be more reliable than parsing process id from the logs, which you stated doesn't work with older versions of VSTest.

Reply:
https://github.com/microsoft/vstest/blob/main/playground/TestPlatform.Playground/Program.cs#L252 there you probably want to implement ITestHostLauncher3, that also gives you info about tfm, and the dlls you will debug. 

This is interesting for multi target framework runs, where the debugger might differ between .net framework and .net.

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/ITestHostLauncher3.cs

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/AttachDebuggerInfo.cs#L8

Reply:
Thanks for the help!

`ITestHostLauncher3` was the solution to my debugging challenges.
`RunTestsWithCustomTestHost` only has overloads taking `ITestHostLauncher`, so I overlooked ITestHostLauncher2 and ITestHostLauncher3. Poking around the TestPlatform code, it looks like there's some casting going on.

Reply:
Also, the TestPlatform playground has been indispensable as I've tried to learn how to work with VSTest.
I'm grateful to the team for open sourcing it!

I extracted a [standalone copy of the TestPlatform Playground](https://github.com/farlee2121/TestPlatform.Playground). That's where I run most of my experiments